File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/reactiveobject/UaReactiveObjectCallback.java
package org.opcfoundation.webserver.addressspace.reactiveobject;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.opcfoundation.webserver.service.message.reactiveobject.*;

import java.util.concurrent.CompletableFuture;

public interface UaReactiveObjectCallback {
    // Browse object child
    default CompletableFuture<BrowseObjectResponse> onBrowseObjectChildren(BrowseObjectRequest request)
    {
        return CompletableFuture.supplyAsync(()-> {
            throw new UaRuntimeException(StatusCodes.Bad_NotImplemented);
        });
    }

    default CompletableFuture<BrowseMemberResponse> onBrowseMemberChildren(BrowseMemberRequest request)
    {
        return CompletableFuture.supplyAsync(()-> {
            throw new UaRuntimeException(StatusCodes.Bad_NotImplemented);
        });
    }

    // Browse object parent
    default CompletableFuture<BrowseObjectResponse> onBrowseObjectParent(BrowseObjectRequest request)
    {
        return CompletableFuture.supplyAsync(()-> {
            throw new UaRuntimeException(StatusCodes.Bad_NotImplemented);
        });
    }

    // Browse object forward links
    default CompletableFuture<BrowseObjectResponse> onBrowseObjectLinks(BrowseObjectRequest request)
    {
        return CompletableFuture.supplyAsync(()-> {
            throw new UaRuntimeException(StatusCodes.Bad_NotImplemented);
        });
    }

    // Read object attributes
    default CompletableFuture<ReadObjectAttributeResponse> onReadObjectAttributes(ReadObjectAttributeRequest request)
    {
        return CompletableFuture.supplyAsync(()-> {
            throw new UaRuntimeException(StatusCodes.Bad_NotImplemented);
        });
    }

    // Read object member attributes
    default CompletableFuture<ReadMemberAttributeResponse> onReadMemberAttributes(ReadMemberAttributeRequest request)
    {
        return CompletableFuture.supplyAsync(()-> {
            throw new UaRuntimeException(StatusCodes.Bad_NotImplemented);
        });
    }

    // Read values of member variables
    default CompletableFuture<ReadVariableValueResponse> onReadVariablesValue(ReadVariableValueRequest request)
    {
        return CompletableFuture.supplyAsync(()-> {
            throw new UaRuntimeException(StatusCodes.Bad_NotImplemented);
        });
    }

    // Write values of member variables
    default CompletableFuture<WriteVariableValueResponse> onWriteVariablesValue(WriteVariableValueRequest request)
    {
        return CompletableFuture.supplyAsync(()-> {
            throw new UaRuntimeException(StatusCodes.Bad_NotImplemented);
        });
    }

    // Call method of object
    default CompletableFuture<MethodCallResponse> onMethodCall(MethodCallRequest request)
    {
        return CompletableFuture.supplyAsync(()->{
            throw new UaRuntimeException(StatusCodes.Bad_NotImplemented);
        });
    }

    // History data read of object
    default CompletableFuture<ReadHistoryDataResponse> onReadHistoryData(ReadHistoryDataRequest request)
    {
        return CompletableFuture.supplyAsync(()->{
            throw new UaRuntimeException(StatusCodes.Bad_NotImplemented);
        });
    }

    // History event read of object
    default CompletableFuture<ReadHistoryEventResponse> onReadHistoryEvent(ReadHistoryEventRequest request)
    {
        return CompletableFuture.supplyAsync(()->{
            throw new UaRuntimeException(StatusCodes.Bad_NotImplemented);
        });
    }
}

File: opcua-webservice-node/source/src/server/addressspace/reactiveobject/UaReactiveObjectCallback.ts
import {
    BrowseMemberRequest,
    BrowseMemberResponse,
    BrowseObjectRequest,
    BrowseObjectResponse,
    MethodCallRequest,
    MethodCallResponse,
    ReadHistoryDataRequest,
    ReadHistoryDataResponse,
    ReadHistoryEventRequest,
    ReadHistoryEventResponse,
    ReadMemberAttributeRequest,
    ReadMemberAttributeResponse,
    ReadObjectAttributeRequest,
    ReadObjectAttributeResponse,
    ReadVariableValueRequest,
    ReadVariableValueResponse,
    WriteVariableValueRequest,
    WriteVariableValueResponse,
    AddObjectRequest,
    AddObjectResponse,
    DeleteObjectRequest,
    DeleteObjectResponse,
    AddReferenceRequest,
    AddReferenceResponse,
    DeleteReferenceRequest,
    DeleteReferenceResponse,
    WriteObjectAttributeRequest,
    WriteObjectAttributeResponse,
} from "../../service/message";

export interface UaReactiveObjectCallback {
    onBrowseObject(request: BrowseObjectRequest): Promise<BrowseObjectResponse>;
    onBrowseMember(request: BrowseMemberRequest): Promise<BrowseMemberResponse>;
    onReadObjectAttributes(request: ReadObjectAttributeRequest): Promise<ReadObjectAttributeResponse>;
    onReadMemberAttributes(request: ReadMemberAttributeRequest): Promise<ReadMemberAttributeResponse>;
    onReadVariablesValue(request: ReadVariableValueRequest): Promise<ReadVariableValueResponse>;
    onWriteVariablesValue(request: WriteVariableValueRequest): Promise<WriteVariableValueResponse>;
    onWriteObjectAttributes(request: WriteObjectAttributeRequest): Promise<WriteObjectAttributeResponse>;
    onMethodCall(request: MethodCallRequest): Promise<MethodCallResponse>;
    onReadHistoryData(request: ReadHistoryDataRequest): Promise<ReadHistoryDataResponse>;
    onReadHistoryEvent(request: ReadHistoryEventRequest): Promise<ReadHistoryEventResponse>;
    onAddObject(request: AddObjectRequest): Promise<AddObjectResponse>;
    onDeleteObject(request: DeleteObjectRequest): Promise<DeleteObjectResponse>;
    onAddReference(request: AddReferenceRequest): Promise<AddReferenceResponse>;
    onDeleteReference(request: DeleteReferenceRequest): Promise<DeleteReferenceResponse>;
}
